Support zooming and panning in the image viewer

MostlyKIGuess opened the pull request:
Implemented Pan and Zoom on the Image Viewer.

Demo:


https://github.com/user-attachments/assets/855bafe8-fdc2-4945-9bfb-e48382264806





Closes #9584



Release Notes:

- Add zoom in/out, reset, fit-to-view and zoom-to-actual actions 
- Support scroll-wheel zoom with modifier, click-and-drag panning 
- Renders a zoom percentage overlay. 
- ImageView Toolbar Added to control zoom in/out and fit view.

MostlyKIGuess committed:
feat: add interactive image viewer with zoom and pan

- add zoom in/out, reset, fit-to-view and zoom-to-actual actions -
support scroll-wheel zoom with modifier, click-and-drag panning -
renders a zoom percentage overlay. - shortcuts keys for all zoom
actions.

MostlyKIGuess committed:
fix: image_viewer in tests

MrSubidubi reviewed (changes requested):
Thanks for this! I'll give this a more proper review soonish (got some other things on my list first), but just to give you a few TODOs I just noticed: 

- Can we add buttons for the new actions to the toolbar, optimally the right corner? If we do not have icons for some of these, we usually source ours from https://lucide.dev/icons/ (Please only do if you actually have to, we should have quite a few already).
- Whatever the checkered background is doing in your example is.. not optimal. Could you ensure that the checkered background is actually just drawn behind the image (so one can see the transparent parts of the image) and that the squares properly follow the scaling of the image?


Thanks!

MostlyKIGuess committed:
feat: add ImageView toolbar with controls

- Refactor checkerboard rendering to respect the scaled image size and
pan - Add ImageViewToolbarControls that observes the active ImageView
and provides zoom in/out, reset

MostlyKIGuess committed:
use theme background instead of checkerboard

MostlyKIGuess commented:
The last commit vastly improved the performance, realized  we didn't even need to draw it :D

MostlyKIGuess commented:
@MrSubidubi Could you review this?

MrSubidubi reviewed (changes requested):
Thanks for the follow up! A few notes:

- We do need the checkered background for partially transparent images, e.g. `zed/crates/zed/resources/app-icon-preview@2x.png`. We do have the outline, but I'd definitely prefer the checkered background, as removing it would feel like a regression.
- Dragging does not account for the current zoom factor and can thus get out of sync. Can add a video of this if you'd like, but this basically means that the image does not move as many pixels as I drag on certain zoom levels. IMO this should be 1:1 mapping without any drift.
- The zoom level in the bottom half of the screen can get out of the view. I think we should probably just remove it, because we have the appropiate zoom level now present in the toolbar as well.

That should be it for this round!

Also, quick note, no need to ping me here. I'll get back to you as soon as I can, but this can (sadly) sometimes take some time. Thanks!

MostlyKIGuess committed:
render board background, removed bottom-right percentage overlay

MostlyKIGuess committed:
Merge branch 'main' into zoom-image

MostlyKIGuess committed:
image with absolute position and pan

- compute left/top from container bounds and scaled size and use
  absolute positioning with pan offsets;
- fall back to flex centering when not positioning manually.
- only clear dragging state and notify when dragging was active.

MostlyKIGuess committed:
simplify paint for faster rendering

- now using 48.0 * zoom for square size, compute rows/cols from bounds,
  skip alternate squares
- clamp tile size to bounds and paint a single grey color.

MostlyKIGuess committed:
light background for similar effect

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into zoom-image

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into zoom-image

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into zoom-image

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into zoom-image

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into zoom-image

MrSubidubi reviewed (changes requested):
Nice, this is so much better already, thank you very much! (And sorry for the delay from my side, was too occupied elsewhere).

I left quite a few comments below, but we should already be close to merging.

Most notably though, we still have two mayor issues as shown in the video here: 

https://github.com/user-attachments/assets/0f555d04-d3f4-4dcc-bb59-de7dea76ef76

As you see,
- there is a brief flicker at the start. This happens I believe because we flex layout once, and I think we should drop the special case for this here. This might be a bit tricky, but we could add a custom element for the image to draw it (would introduce a bit of boilerplate), but I'd like us to get rid of this. If you need more input for this, please hit me up and I'll get into more detail here/look at this myself.
- zooming with the mouse wheel does not take the current positon of the mouse into account and that feels counterintuitive. I think when we resize there, the resize-operation should take the cursor position into account.

Other than that, looks so much better, thank you already! As for profiling, we are currently doing and experimenting a lot with tracy, but there is no official guideline. So far performance has been fine for me, although I fully agree checkered background support would be nice to have here. Feel free to throw a TODO somewhere, but we can also discuss more here.

MostlyKIGuess committed:
refactor(image_viewer): implement mouse-relative zoom and fix layout
flicker

- added basic info on actions and extracted magic numbers to constants
- use intuitive scaling with translation of (1-scale_factor) on the
  offset
- attach global listener only during drag

MostlyKIGuess commented on crates/image_viewer/src/image_viewer.rs:
```diff
@@ -303,81 +474,191 @@ impl Focusable for ImageView {
+                    },
+                )
+                .absolute()
+                .size_0()
```
Yes!

MostlyKIGuess commented:
Thank you for the detailed review! 

> Other than that, looks so much better, thank you already! As for profiling, we are currently doing and experimenting a lot with tracy, but there is no official guideline. So far performance has been fine for me, although I fully agree checkered background support would be nice to have here. Feel free to throw a TODO somewhere, but we can also discuss more here.

Yeah, this is acceptable performance compared to earlier.
Maybe this goes into GPUI instead of it being here. So the follows ups would be:
1. To have background support from GPUI. 
2. I also wanted to add that I have a [PR](https://github.com/gpui-ce/gpui-ce/pull/11) in the gpui-ce for touchpad support in linux and mac.  Perhaps that can be added later for zooming in through the touchpad.

MrSubidubi reviewed (changes requested):
Okay, one mast mid-sized thing and I think we are good, thank you for all your efforts.

I think we need to refactor this a bit so that we have the container bounds of the current frame and not the bounds of the previous frame. Using the information from the previous frame, we
- always do not render an image on the first frame, which causes a brief but noticeable flicker
- always have a small flicker whenever we open other panels/resize because again, sizing calculation is done with sizes from the previous frame.

I think we need to change stuff of this to utilize a small wrapper instead that impls `Element`. We can then request the layout as we currently do and get the corresponding bounds during prepaint, which we can then utilize and which corresponds to the actual current bounds as opposed to the bounds from the previous frame. Then we should basically be able to use the current `render` of the image viewer in that prepaint, request the layout slightly delayed there and can prepaint and paint after as needed.

I think https://github.com/zed-industries/zed/blob/81ea0403d1b9e94f0b3d1b7e2bb4857a0adfc060/crates/ui/src/utils/with_rem_size.rs#L45 might be a good skeleton for this in combination with `StickyItems` (obviously we'd need this to be slightly different), but this would resolve that issue. What do you think?

MostlyKIGuess commented:
> I think that is precisely what I wanted to describe and I think we cannot get rid of this unless going for an `Element` impl - I am happy to stand corrected here, but I think that we will not get around this if we want to update/paint the image on the same frame.

Yeah, this can be added as a todo on GPUI perhaps where the render is parallel for the background and the image. To describe it precisely an ability to set an image background/pattern.

MostlyKIGuess committed:
refactor(image_viewer): use Element wrapper

- Replace canvas-based container bounds tracking with
  ImageContentElement that wraps a Div and updates bounds during
  prepaint.
- This ensures positioning uses current-frame bounds instead of
  previous-frame bounds.

MrSubidubi committed:
Ensure container bounds are known on first render

MostlyKIGuess commented:
> Went ahead and applied the last set of fixes myself - feel free to check these out, we should now always properly layout the image according to the current viewport, no matter what. If there is anything, please let me know, otherwise, I think we are good to go!

Thank you! I tested it locally and works as expected. 

I can add gesture support along with adding the following changes in GPUI: [Gesture Support GPUI-CE PR](https://github.com/gpui-ce/gpui-ce/pull/11) as continuation.

MrSubidubi reviewed (approved):
Thanks for taking another look!

I think the pinch support sounds interesting, feel free to open a PR for that with the support for the image viewer here and I'll gladly take a look!


Lastly, thank you so so much for bearing with me here and for your perseverance, I for one think (and hope you agree) it was worthwhile and am very happy to merge this now. Thank you so much!

MostlyKIGuess commented:
> Lastly, thank you so so much for bearing with me here and for your perseverance, I for one think (and hope you agree) it was worthwhile and am very happy to merge this now. Thank you so much!

Thanks a ton for the thorough review and patience, really appreciate it. Learned a lot from this one. 

Looking forward to the pinch-zoom follow-up: https://github.com/zed-industries/zed/pull/47351

MrSubidubi commented:
Hey @MostlyKIGuess ! Since this landed, we now have checkered backgruound support in our shaders! 🎉  Can I interest you in looking into that? 

Happy to look at a PR, I'll probably do it myself at some point but currently don't have the time to do it immediately :/
